src/Features/Core/Portable/ExtractClass/ExtractClassWithDialogCodeAction.cs

```diff
@@ -0,0 +1,260 @@
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var typeDeclaration = root.GetAnnotatedNodes(symbolMapping.TypeNodeAnnotation).SingleOrDefault();
+                if (typeDeclaration == null)
+                {
+                    continue;
+                }
+
+                foreach (var memberAnalysis in remainingResults.ToArray())
+                {
+                    var annotation = symbolMapping.SymbolToDeclarationAnnotationMap[memberAnalysis.Member];
+
+                    var nodeOrToken = typeDeclaration.GetAnnotatedNodesAndTokens(annotation).SingleOrDefault();
+                    var node = nodeOrToken.IsNode
+                        ? nodeOrToken.AsNode()
+                        : nodeOrToken.AsToken().Parent;
+
+                    if (node is null)
+                    {
+                        continue;
+                    }
+
+                    var currentSymbol = semanticModel.GetDeclaredSymbol(node, cancellationToken);
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    pullMembersBuilder.Add((currentSymbol, memberAnalysis.MakeAbstract));
+                    remainingResults.Remove(memberAnalysis);
+                }
+            }
+
+            Contract.ThrowIfFalse(remainingResults.Count == 0);
+
+            var pullMemberUpOptions = PullMembersUpOptionsBuilder.BuildPullMembersUpOptions(newType, pullMembersBuilder.ToImmutable());
+            var updatedOriginalDocument = solution.GetRequiredDocument(_document.Id);
+
+            return await MembersPuller.PullMembersUpAsync(updatedOriginalDocument, pullMemberUpOptions, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<INamedTypeSymbol> GetNewTypeSymbolAsync(Document document, INamedTypeSymbol typeToBeLookedFor, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            return (INamedTypeSymbol)semanticModel.Compilation.GetSymbolsWithName(typeToBeLookedFor.Name, SymbolFilter.Type, cancellationToken).Single();
+        }
+
+        private static async Task<Solution> GetSolutionWithBaseAddedAsync(
+            Solution solution,
+            AnnotatedSymbolMapping symbolMapping,
+            INamedTypeSymbol newType,
+            ImmutableArray<ExtractClassMemberAnalysisResult> memberAnalysisResults,
+            CancellationToken cancellationToken)
+        {
+            var unformattedSolution = solution;
+            var remainingResults = new List<ExtractClassMemberAnalysisResult>(memberAnalysisResults);
+
+            foreach (var documentId in symbolMapping.DocumentIds)
```

Comment:
why do we have multiple IDs?  why would we not just update one single doc in one single location.  namely the base list of the type we selected (or the containing type of the member we selected).

Reply:
The type could be abstract and across multiple documents. We don't restrict the members from other documents from being included in the items being pulled up. I'll add a test for that since it's missing.